### DISCUSSION TITLE
flash adapter for V3

### DISCUSSION THREAD

**Discussion by Chomemel** [sanni/cartreader#538] [General Chat]
Hi,
I don't if I can ask it here. If you feel it's not appropiate, please delete the post.

I want to order a flash adapter (finished with ENIG), but you can't order less than 5 PCB. I was wondering if someone in Europe (France would be even better) already ordered some and had some PCB left.

**Comment by Chomemel** [comment]
PCB ordered, pins soldered, but I totally forgot the switches.

Can someone confirm that [these ones](https://fr.aliexpress.com/item/32963237441.html?spm=a2g0o.productlist.main.1.42df16c1OzBrHR&algo_pvid=20ab7737-57d0-44cd-a2ed-fda2b96b4215&algo_exp_id=20ab7737-57d0-44cd-a2ed-fda2b96b4215-0&pdp_ext_f=%7B"sku_id"%3A"66610156742"%7D&pdp_npi=2%40dis%21EUR%212.55%211.53%21%21%21%21%21%402100b20d16749865962912166d06ef%2166610156742%21sea&curPageLogUid=ZhYWhGm6Ms8y) would fit. It seems they should, but it's just to be 100% sure.

**Reply by sanni** [reply]
Yes those will fit.

Measurements taken from PCB vs. description of switch from Aliexpress, close enough:
![smd_switch](https://user-images.githubusercontent.com/19169494/215344696-1c9f56ce-d6a0-40c2-be86-c9b41f29e9b6.jpg)

**Reply by Chomemel** [reply]
Thanks Sanni for your answer. Quick as usual.

**Reply by Chomemel** [reply]
Switches soldered. Programming 3211 was a success.

**Comment by Chomemel** [comment]
I've been able to flash 29LV640 and 29LV320 so far, via the TSOP48 adapter.

I saw SOP44 29LV160 are compatible with the CR. What about the TSOP48 version of the 29LV160? Could it work using the TSOP48 adapter?

**Reply by sanni** [reply]
I have not tested it but it should work as it has the same pinout as the 640 variant just missing a few address lines.

**Reply by Chomemel** [reply]
Yes, also same pinout as the 320. I'll give it a try as soon as I can. Thanks for the answer.

**Reply by Chomemel** [reply]
So I tried and it turns out, my 29LV160 are AM29LV160, not MX... So the ID is good (0149). But I can't do anything with it. It goes to the "First 40 bytes" when I press a button, then go back to main menu. I'm using v9.1.
Should I upgrade to a more recent version?
Or do I have to add lines in the code? 
![IMG_20230228_001138](https://user-images.githubusercontent.com/107198559/221713042-db25e62e-1021-4205-88d6-3b333b553eb4.jpg)
![IMG_20230228_001207](https://user-images.githubusercontent.com/107198559/221713098-b55bb56c-99b0-4330-a27b-dd36c61b214c.jpg)

**Reply by sanni** [reply]
That means that there is no flash function for the AM29LV160 yet. You need to read the datasheet and then edit flash.ino accordingly.

**Reply by Chomemel** [reply]
I have zero experience in coding and not much knowledges in flashroms, but guess it's time to start learning.
Thanks for the answers.

**Reply by sanni** [reply]
In [line 167](https://github.com/sanni/cartreader/blob/master/Cart_Reader/FLASH.ino#L167) of flash.ino add
````
          else if (flashid == 0x0149)
            writeFlash29LV640();
````

Above [line 422](https://github.com/sanni/cartreader/blob/master/Cart_Reader/FLASH.ino#L422) of flash.ino add
````
  } else if (flashid == 0x0149) {
    println_Msg(F("AM29LV160 detected"));
    println_Msg(F("ATTENTION 3.3V"));
    flashSize = 2097152;
    flashromType = 2;
````

Then see how far it goes now.

**Reply by Chomemel** [reply]
I tried uploading the software to 12.3 with the FLASH.ino modified and there are weird things.
- there is not that screen with the flshrom ID and name (like in my first photo). It goes straight to the menu (blankcheck, read, write...)
- can't read or write anything (it goes to the right screen but writing, reading and everything takes 0 second)
- when "finished" and I press a button I go straight to the screen "N64 Cart reader (with submenu read rom, read save...)
All this tests were done with a MX29LV320 and MX29LV3211 that worked before I upgraded the software.

EDIT: I tried to flash the arduino with the software without modifing the flash.ino and it does exactly the same.

EDIT: Maybe it will be easier with a video

https://user-images.githubusercontent.com/107198559/224533605-35d0fd16-f4a1-48d6-9b40-4c3943d130b4.mp4

**Reply by Chomemel** [reply]
Ok, I tried with the v9.1, changing the code like that (the bold part):

>  switch (flashromType) {
        case 1:
          writeFlash29F032();
          break;
        case 2:
          if (strcmp(flashid, "C2F3") == 0)
            writeFlash29F1601();
          else if ((strcmp(flashid, "C2F1") == 0) || (strcmp(flashid, "C2F9") == 0))
            writeFlash29F1610();
          else if ((strcmp(flashid, "C2C4") == 0) || (strcmp(flashid, "C249") == 0) ||
                   (strcmp(flashid, "C2A7") == 0) || (strcmp(flashid, "C2A8") == 0) ||
                   (strcmp(flashid, "C2C9") == 0) || (strcmp(flashid, "C2CB") == 0) ||
                   **(strcmp(flashid, "0149") == 0) || (strcmp(flashid, "01C4") == 0))**

>   case 3:
      filePath[0] = '\0';
      sd.chdir("/");
      fileBrowser(F("Select file"));
      display_Clear();
      time = millis();
      if (strcmp(flashid, "C2F3") == 0) {
        writeFlash16_29F1601();
      }
      else if ((strcmp(flashid, "C2C4") == 0) || (strcmp(flashid, "C249") == 0) || (strcmp(flashid, "C2A7") == 0) || (strcmp(flashid, "C2A8") == 0) || (strcmp(flashid, "C2C9") == 0) || (strcmp(flashid, "C2CB") == 0) || (strcmp(flashid, "C2FC") == 0) || **(strcmp(flashid, "0149") == 0) || (strcmp(flashid, "01C4") == 0)**) {
        writeFlash16_29LV640();

> **else if ((strcmp(flashid, "0149") == 0) || (strcmp(flashid, "01C4") == 0)) {
    println_Msg(F("AM29LV160 detected"));
    println_Msg(F("ATTENTION 3.3V"));
    flashSize = 2097152;
    flashromType = 2**;

So far it works. I tried to blankcheck, write, read.

Can someone tell me if what I add in the code seemes correct? 

If everything works fine with v9.1, what could be the problem with 12.3?

**Comment by Chomemel** [comment]
Another question. Maybe it's a dumb one.
The Cart Reader is able read/earse/write 29LV... from macronix manufacturer (MX). What about the 29LV... from other manufacturers? For example AM29LV... or EN29LV... ? Are they recognised as MX? Are they recognised at all?
I took the exemple of MX29LV..., but it could have been for antother type of EEPROM.

**Reply by sanni** [reply]
I only tried with MX and used MX datasheet but it could be that other 29LV chips behave exactly the same. It could also be that they need their flash id added though.

**Reply by Chomemel** [reply]
Ok, thanks. To be honest I only have MX29LV... (I'll have to check). If I have EEPROM from other manufacturers than the ones in the code, I'll give it a try.

**Comment by Chomemel** [comment]
I'd like to add some EEPROMs for v12.5 or later (mostly 29LVxxx) . I tried on v9.1 and it works. I'll try them on the last version when I find some time. Is there a way I can add the lines to the public code or send them to an administator to add it?

**Reply by sanni** [reply]
On https://github.com/sanni/cartreader you click on Fork in the top right corner, this will create your own Chomemel/cartreader github. There you can upload your changed files. When all is ready you can create a pull request which will transfer the changes to the sanni/cartreader github. At least that how I think it goes, never done it myself.

**Reply by Chomemel** [reply]
I pulled a request. But... 
I downloaded the last version of flash.ino, then I did the same changes I did in v9.1.
Then I flashed my arduino with v12.4, replacing the flash.ino with the last one with my changes.

It turns out in v12.4 I couldn't find the "FLASH" in the modules like in v9.1. The only one I can find is "flash rom programmer for SNES repro".
That could explain why I can't find the same menus in the flash rom menu. Did I miss something?

**Reply by Ancyker** [reply]
As far as I understand it, `enable_FLASH` is required for the SNES repros, and `enable_FLASH16` is required for full flash support. Someone who actually uses that module could likely clarify further.

**Reply by Chomemel** [reply]
Excellent understanding !!!
And now that you mentionned it, I can see it in the FLASH.ino
> #ifdef enable_FLASH16
// Flash start menu
static const char flashMenuItem1[] PROGMEM = "8bit Flash adapter";
static const char flashMenuItem2[] PROGMEM = "Eprom adapter";
static const char flashMenuItem3[] PROGMEM = "16bit Flash adapter";

The menu reappered
![IMG_20230514_124542](https://github.com/sanni/cartreader/assets/107198559/08b43d2a-ea9e-419b-9a01-9ee086f14ad4)

And my modified flash.ino works great
![IMG_20230514_124601](https://github.com/sanni/cartreader/assets/107198559/901836b2-9c8d-4b3d-aadc-868d32f6c45c)

AM29LV160 is detected. I could read, erase, blankcheck and flash it (like I did with v9.1)


**So the flash.ino I pulled a request for seems to work just fine.**

I take a minute to thank you Ancyker (and also the other contibutors) for your amazing work lately.

**Comment by sanni** [comment]
> I pulled a request. 
> 
> It turns out in v12.4 I couldn't find the "FLASH" in the modules like in v9.1.


I have just added your pull request.

[12.4 is bugged](https://github.com/sanni/cartreader/issues/777) when it comes to flashroms you should use the latest [github code](https://github.com/sanni/cartreader/archive/refs/heads/master.zip). 

> As far as I understand it, `enable_FLASH` is required for the SNES repros, and `enable_FLASH16` is required for full flash support. Someone who actually uses that module could likely clarify further.

[enable_FLASH](https://github.com/sanni/cartreader/blob/master/Cart_Reader/Config.h#L107) is for flashing SNES repros and for the Flashrom Programmer menu option. They both share the same code because the [flash_adapter PCB](https://github.com/sanni/cartreader/wiki/Flash-Adapter) is loosely based on the SNES hirom PCB design. The flash_adapter PCB lets you flash single flashrom chips. This turns the Cart Reader into a more general flashrom programmer e.g. for programming a Dreamcast bios chip. It is enabled by default.

[enable_FLASH16](https://github.com/sanni/cartreader/blob/master/Cart_Reader/Config.h#L108) is an extension to enable_FLASH and unlocks the "Eprom adapter" and "16bit Flash adapter" sub-menus under the Flashrom Programmer menu. 
You should not have to use enable_FLASH16 unless you got the 26L6420_adapter or [27c322_adapter PCBs or the super old 16bit flashrom adapter PCB](https://github.com/sanni/cartreader/discussions/754#discussioncomment-5501101), so flash adapter PCBs that use the two extra flaps left and right of the SNES cart slot. That's why it is disabled by default.

**Reply by Chomemel** [reply]
Thanks Sanni for the precisions.


> 12.4 is bugged when it comes to flashroms you should use the latest [github code](https://github.com/sanni/cartreader/archive/refs/heads/master.zip).*

Yeah, that's what I did. I downloaded the last version of flash.ino, modified it, copied it into 12.4 sketckbooks, and flashed my arduino. And it worked.

Also, another question. When I want to flash an EEPROM, the folder \flash appears under every txt files. Is there a way to put it under the other folders?

**Reply by sanni** [reply]
Set the database txt files to hidden in Windows. See [here](https://github.com/sanni/cartreader/wiki/Preparing-the-SD-card)

**Reply by Chomemel** [reply]
Neat !

I see a lot of the wiki has been updated. So much has been done in one year (when I discovered the Cart Reader).
Thanks again to everyone for the good work.

**Reply by Chomemel** [reply]
As the code is changed, I'd like to change the [MX29LV320 wiki page](https://github.com/sanni/cartreader/wiki/MX29LV320) to make a more general page, including other TSOP48 that can be flashed like MX29LV320 (like 29LV160, 29LV640).
If I click on "EDIT", will it edit the page directly? Or will it "pull a request" to change the text?
